### PR TITLE
replace address decoder of XBar to Espresso decoder, also refactor Parameters.scala to chisel3

### PR DESCRIPTION
**Pull request by Lucas-Wye**

### CODE REVIEW

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (jerryz123) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:18]: Use 2-space indents, not tabs

File: src/main/scala/tilelink/Xbar.scala
Comment (jerryz123) [inline src/main/scala/tilelink/Xbar.scala:36]: What does this comment mean? Also why does arbiter choice affect decoder

File: src/main/scala/tilelink/Xbar.scala
Comment (jerryz123) [inline src/main/scala/tilelink/Xbar.scala:206]: These comments should not be inline in the final commit.

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (Lucas-Wye) [reply src/main/scala/diplomacy/EspressoAddressDecoder.scala]: Thank you, I have fixed it.

File: src/main/scala/tilelink/Xbar.scala
Comment (Lucas-Wye) [reply src/main/scala/tilelink/Xbar.scala]: yes, these are some code from another branch, I just removed them, thanks

File: src/main/scala/tilelink/Xbar.scala
Comment (Lucas-Wye) [reply src/main/scala/tilelink/Xbar.scala]: has been removed

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:18]: I think it needs more comments.

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:27]: same here...

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:40]: IIRC
```suggestion
    val dontCarePat: BitPat = BitPat.dontCare(maxBits)
```

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:41]: `take` here scared me, in which case, the decoder width won't match `ports` width

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:46]: nitpick: add public api to `AddressSet`: `toBitPat(xLen: Int)` might be better, if not, make this function private.

**Review by sequencer — CHANGES_REQUESTED** [review]

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (sequencer) [inline src/main/scala/diplomacy/EspressoAddressDecoder.scala:10]: I think it should be the name `EspressoAddressDecoder`: if user not installing `espresso`, this is gone be `QMCDecoder` from chisel

Comment (sequencer) [reply src/main/scala/diplomacy/EspressoAddressDecoder.scala:10]: I think you can move logic in `EspressoAddressDecoder` to `AddressDecoder`, since after this PR is merged, original `AddressDecoder` is going to be dead code.

**Comment by sequencer** [comment]
FYI:
Here are some examples to use chisel decoder as `AddressDecoder`:
Example from @jerryz123 
https://github.com/ucb-bar/constellation/blob/216eee92be30a8d24a3e3369b6f23c6f43b0d3a3/src/main/scala/router/RouteComputer.scala#L29
Example from @sequencer 
https://github.com/chipsalliance/chisel3/blob/2b48fd15a7711dcd44334fbbc538667a102a581a/src/test/scala/chiselTests/util/BitSetSpec.scala#L90-L116

**Comment by sequencer** [comment]
`decoder.bitset` provide a relatively clear API for `AddressSet`. Basically an `AddressSet` is a `BitSet`(but with unknown width IIRC), so if you can use `decoder.bitset` API, I bet it will be much clear.

File: src/main/scala/diplomacy/EspressoAddressDecoder.scala
Comment (Lucas-Wye) [reply src/main/scala/diplomacy/EspressoAddressDecoder.scala:41]: Because the decoder width is determined by maxBits, which is the maximum number of bit in all input address. As a result, the decoder width is usually larger than port width.

**Comment by sequencer** [comment]
This is blocked by chipsalliance/chisel3#2883, we should merge this into `develop` branch

**Comment by sequencer** [comment]
wait for chipsalliance/tilelink

**Pull request closed without merging**
